egui/src/input_state/touch_state.rs

```diff
@@ -0,0 +1,273 @@
+        self.active_touches.remove(&id);
+        // for now we only support exactly two fingers:
+        if self.active_touches.len() == 2 {
+            self.start_gesture(time);
+        } else {
+            self.end_gesture()
+        }
+    }
+
+    fn start_gesture(&mut self, time: f64) {
+        for mut touch in self.active_touches.values_mut() {
+            touch.gesture_start_pos = touch.pos;
+        }
+        if let Some((touch1, touch2)) = self.both_touches()
+        // always true
+        {
+            let start_dyn_state = DynGestureState {
+                pos: self::center_pos(touch1.pos, touch2.pos),
+                distance: self::distance(touch1.pos, touch2.pos),
+                direction: self::direction(touch1.pos, touch2.pos),
+                force: (touch1.force + touch2.force) * 0.5,
+            };
+            self.gesture_state = Some(GestureState {
+                start_time: time,
+                start: start_dyn_state,
+                previous: start_dyn_state,
+                current: start_dyn_state,
+            });
+        }
+    }
+
+    fn update_gesture(&mut self) {
+        if let Some((touch1, touch2)) = self.both_touches() {
+            let state_new = DynGestureState {
+                pos: self::center_pos(touch1.pos, touch2.pos),
+                distance: self::distance(touch1.pos, touch2.pos),
+                direction: self::direction(touch1.pos, touch2.pos),
+                force: (touch1.force + touch2.force) * 0.5,
+            };
+            if let Some(ref mut state) = &mut self.gesture_state
+            // always true
+            {
+                state.previous = state.current;
+                state.current = state_new;
+            }
+        }
+    }
+
+    fn end_gesture(&mut self) {
+        self.gesture_state = None;
+    }
+
+    fn both_touches(&self) -> Option<(&ActiveTouch, &ActiveTouch)> {
+        if self.active_touches.len() == 2 {
+            let mut touches = self.active_touches.values();
+            let touch1 = touches.next().unwrap();
+            let touch2 = touches.next().unwrap();
+            Some((touch1, touch2))
+        } else {
+            None
+        }
+    }
+}
+
+impl TouchState {
+    pub fn ui(&self, ui: &mut crate::Ui) {
+        ui.label(format!("{:?}", self));
+    }
+}
+
+impl Debug for TouchState {
+    // We could just use `#[derive(Debug)]`, but the implementation below produces a less cluttered
+    // output:
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for (id, touch) in self.active_touches.iter() {
+            f.write_fmt(format_args!("#{:?}: {:#?}\n", id, touch))?;
+        }
+        f.write_fmt(format_args!("gesture: {:#?}\n", self.gesture_state))?;
+        Ok(())
+    }
+}
+
+fn center_pos(pos_1: Pos2, pos_2: Pos2) -> Pos2 {
+    pos2((pos_1.x + pos_2.x) * 0.5, (pos_1.y + pos_2.y) * 0.5)
+}
+
+fn distance(pos_1: Pos2, pos_2: Pos2) -> f32 {
+    (pos_2 - pos_1).length()
+}
```

Comment:
you can already do `pos_1.distance(pos_2)`

Reply:
handy, thanks. The `distance` function has now dissolved into a new fn `calc_averages()`